File: frontend/src/types/inventory.ts
import type { Product, Warehouse } from "./catalog";

export type StockStatus = "NORMAL" | "LOW_STOCK" | "OUT_OF_STOCK";

export type StockMovementType =
  | "OPENING_BALANCE"
  | "GOODS_RECEIPT"
  | "STOCK_ISSUE"
  | "TRANSFER_OUT"
  | "TRANSFER_IN"
  | "ADJUSTMENT_INCREASE"
  | "ADJUSTMENT_DECREASE"
  | "RETURN";

export interface InventoryItem {
  product_id: number;
  sku: string;
  product_name: string;
  category_name: string;
  preferred_supplier_name: string | null;
  unit_of_measure: string;
  reorder_level: number;
  total_on_hand: number;
  total_reserved: number;
  available_quantity: number;
  inventory_value: string;
  stock_status: StockStatus;
  warehouse_count: number;
}

export interface LowStockItem extends InventoryItem {
  suggested_reorder_quantity: number;
}

export interface InventoryBalance {
  id: number;
  warehouse: Warehouse;
  product_id: number;
  quantity_on_hand: number;
  quantity_reserved: number;
  available_quantity: number;
  inventory_value: string;
  stock_status: StockStatus;
  updated_at: string;
}

export interface StockMovement {
  id: number;
  product_id: number;
  product_sku: string;
  product_name: string;
  warehouse_id: number;
  warehouse_code: string;
  warehouse_name: string;
  movement_type: StockMovementType;
  quantity: number;
  reference_type: string | null;
  reference_id: string | null;
  reason: string;
  performed_by: number | null;
  performed_by_name: string | null;
  created_at: string;
}

export interface ProductInventoryDetail {
  product: Product;
  total_on_hand: number;
  total_reserved: number;
  available_quantity: number;
  inventory_value: string;
  stock_status: StockStatus;
  balances: InventoryBalance[];
  recent_movements: StockMovement[];
}

File: backend/app/schemas/inventory.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.models.inventory import StockMovementType
from app.schemas.product import ProductRead
from app.schemas.warehouse import WarehouseRead


class InventoryBalanceRead(BaseModel):
    id: int
    warehouse: WarehouseRead
    product_id: int
    quantity_on_hand: int
    quantity_reserved: int
    available_quantity: int
    inventory_value: Decimal
    stock_status: str
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class InventoryItem(BaseModel):
    product_id: int
    sku: str
    product_name: str
    category_name: str
    preferred_supplier_name: str | None
    unit_of_measure: str
    reorder_level: int
    total_on_hand: int
    total_reserved: int
    available_quantity: int
    inventory_value: Decimal
    stock_status: str
    warehouse_count: int


class LowStockItem(InventoryItem):
    suggested_reorder_quantity: int
    suggested_target_quantity: int


class StockMovementRead(BaseModel):
    id: int
    product_id: int
    product_sku: str
    product_name: str
    warehouse_id: int
    warehouse_code: str
    warehouse_name: str
    movement_type: StockMovementType
    quantity: int
    reference_type: str | None
    reference_id: str | None
    reason: str
    performed_by: int | None
    performed_by_name: str | None
    created_at: datetime


class ProductInventoryDetail(BaseModel):
    product: ProductRead
    total_on_hand: int
    total_reserved: int
    available_quantity: int
    inventory_value: Decimal
    stock_status: str
    balances: list[InventoryBalanceRead]
    recent_movements: list[StockMovementRead]
